src/mrack/providers/aws.py

```diff
@@ -209,9 +212,87 @@ async def validate_hosts(self, reqs):
         """Validate that host requirements are well specified."""
         return
 
-    async def can_provision(self, hosts):
-        """Check that hosts can be provisioned."""
-        return True
+    async def get_subnet_available_ips(self, subnet_id, log_msg_start):
+        """Get number of IPs available in a subnet."""
+        try:
+            subnet = self.ec2.Subnet(subnet_id)
+        except ClientError:
+            logger.warning(
+                f"{log_msg_start} Error retrieving info from subnet: {subnet_id}"
+            )
+            return 0
+
+        logger.debug(
+            f"{log_msg_start} Subnet Id: {subnet_id}\n"
+            f"\tHas {subnet.available_ip_address_count} IP addresses available"
+        )
+
+        return subnet.available_ip_address_count
+
+    async def can_provision(self, hosts):  # pylint: disable=too-many-branches
+        """Check that all host can be provisioned.
+
+        Checks:
+        * Available IPv4 addresses are enough
+        """
+        log_msg_start = self.dsp_name
+
+        ip_availabilities = {}  # Dict for storing available IPs in every subnet
+        single_subnet_prov = []  # List for storing provided single subnets
+        mult_subnets_prov = []  # List for storing lists of provided multiple subnets
+
+        # For all hosts, get subnets in dict (with no repetitions) and list of
+        # provided single & multiple subnets
+        for host in hosts:
+            if subnet_ids := host.get("subnet_ids"):
+                if len(subnet_ids) == 1:
+                    single_subnet_prov.append(subnet_ids[0])
+                    if subnet_ids[0] not in ip_availabilities:
+                        ip_availabilities[subnet_ids[0]] = 0
+                else:
+                    mult_subnets_prov.append(subnet_ids)
+                    for subnet_id in subnet_ids:
+                        if subnet_id not in ip_availabilities:
+                            ip_availabilities[subnet_id] = 0
+            else:
+                logger.debug(
+                    f"{log_msg_start} No subnet/s specified for host {host['name']}."
+                )
+
+        # Get available IPs from AWS for every subnet
+        ips_count_wait = [
+            self.get_subnet_available_ips(subnet_id, log_msg_start)
+            for subnet_id in ip_availabilities
+        ]
+        ips_count_result = await asyncio.gather(*ips_count_wait)
+        if not all(ips_count_result) > 0:
```

Comment:
This does not seem like valid logic operation all returns True/False i believe. so enough is to ask for all()
```suggestion
        if not all(ips_count_result):
```

Reply:
Fixed